src/compiler/checker.ts

```diff
@@ -16871,23 +16848,60 @@ namespace ts {
             return isLengthPushOrUnshift || isElementAssignment;
         }
 
-        function maybeTypePredicateCall(node: CallExpression) {
-            const links = getNodeLinks(node);
-            if (links.maybeTypePredicate === undefined) {
-                links.maybeTypePredicate = getMaybeTypePredicate(node);
+        function isDeclarationWithExplicitTypeAnnotation(declaration: Declaration | undefined) {
+            return !!(declaration && (
+                declaration.kind === SyntaxKind.VariableDeclaration || declaration.kind === SyntaxKind.Parameter ||
+                declaration.kind === SyntaxKind.PropertyDeclaration || declaration.kind === SyntaxKind.PropertySignature) &&
+                (declaration as VariableDeclaration | ParameterDeclaration | PropertyDeclaration | PropertySignature).type);
+        }
+
+        function getExplicitTypeOfSymbol(symbol: Symbol) {
+            return symbol.flags & (SymbolFlags.Function | SymbolFlags.Method | SymbolFlags.Class | SymbolFlags.ValueModule) ||
+                symbol.flags & (SymbolFlags.Variable | SymbolFlags.Property) && isDeclarationWithExplicitTypeAnnotation(symbol.valueDeclaration) ?
+                getTypeOfSymbol(symbol) : undefined;
+        }
+
+        function getTypeOfDottedName(node: Expression) {
+            // We require the dotted function name in an assertion expression to be comprised of identifiers
+            // that reference function, method, class or value module symbols; or variable, property or
+            // parameter symbols with declarations that have explicit type annotations. Such references are
+            // resolvable with no possibility of triggering circularities in control flow analysis.
+            switch (node.kind) {

```

Comment:
We should consider handling `ParenthesizedExpression` (both in binding and here) - I can't justify why `(this).assert()` shouldn't assert, other than "just because". It also might be worth renaming to `getDeclaredTypeOfFlowReference` and mimicking the traversals we support in `isMatchingReference`, this way we can know that if we can detect it as a reference, we can assert on it. At least, that kind of consistency would be nice.

Reply:
Agreed.